rundeckapp/src/main/groovy/org/rundeck/app/data/providers/GormTokenDataProvider.groovy

```diff
@@ -161,17 +173,137 @@ class GormTokenDataProvider implements TokenDataProvider {
     }
 
     @Override
+    @GrailsCompileStatic(TypeCheckingMode.SKIP)
     AuthenticationToken tokenLookup(final String token){
-        def authToken = AuthToken.tokenLookup(token);
-        return authToken ?: null
+        def tokenHash = AuthToken.encodeTokenValue(token, AuthTokenMode.SECURED)
+        return AuthToken.createCriteria().get {
+            or {
+                and {
+                    eq("tokenMode", AuthTokenMode.SECURED)
+                    eq("token", tokenHash)
+                }
+                and {
+                    or {
+                        isNull("tokenMode")
+                        eq("tokenMode", AuthTokenMode.LEGACY)
+                    }
+                    eq("token", token)
+                }
+            }
+            or {
+                isNull("type")
+                eq("type", AuthenticationToken.AuthTokenType.USER)
+            }
+        }
+    }
 
+    @Override
+    @GrailsCompileStatic(TypeCheckingMode.SKIP)
+    AuthenticationToken tokenLookupWithType(final String token, AuthenticationToken.AuthTokenType tokenType){
+        def tokenHash = AuthToken.encodeTokenValue(token, AuthTokenMode.SECURED)
+        return AuthToken.createCriteria().get {
+            eq("type", tokenType)
+            or {
+                and {
+                    eq("tokenMode", AuthTokenMode.SECURED)
+                    eq("token", tokenHash)
+                }
+                and {
+                    or {
+                        isNull("tokenMode")
+                        eq("tokenMode", AuthTokenMode.LEGACY)
+                    }
+                    eq("token", token)
+                }
+            }
+        }
     }
 
     @Override
-    AuthenticationToken tokenLookupWithType(final String token, AuthenticationToken.AuthTokenType type){
-        def authToken = AuthToken.tokenLookup(token, type);
-        return authToken ?: null
+    @GrailsCompileStatic(TypeCheckingMode.SKIP)
+    Integer countTokensByUser(String userId) {
+        return AuthToken.createCriteria().count {
+            eq("user", User.get(userId.toLong()))
+            or {
+                eq("type", AuthenticationToken.AuthTokenType.USER)
+                isNull("type")
+            }
+        }
+    }
 
+    @Override
+    @GrailsCompileStatic(TypeCheckingMode.SKIP)
+    Integer countTokensByCreator(String creatorLoginName) {
+        return AuthToken.createCriteria().count {
+            eq("creator", creatorLoginName)
+            or {
+                eq("type", AuthenticationToken.AuthTokenType.USER)
+                isNull("type")
+            }
+        }
+    }
+
+    @Override
+    @GrailsCompileStatic(TypeCheckingMode.SKIP)
+    Integer countTokensByType(AuthenticationToken.AuthTokenType type) {
+        return AuthToken.createCriteria().count {
+            or {
+                eq("type", type)
+                isNull("type")   //type could be null if token was created before we recorded token types
```

Comment:
I think we should remove support for "legacy" user tokens at some point.

Reply:
Agreed, unless someone is upgrading from a very legacy version we shouldn't need to check this. Not sure what the priority would be on removing the extra check.

Reply:
i can backlog it